File: experiments/adrianco/experiment-13/runs/language=go_model=sonnet_prompt=ATDD/rep3/internal/tools/teams.go
package tools

import (
	"context"
	"encoding/json"
	"strings"

	"brazilian-soccer-mcp/internal/data"
	"github.com/mark3labs/mcp-go/mcp"
	"github.com/mark3labs/mcp-go/server"
)

func GetTeamStatsTool(matches []data.Match) server.ServerTool {
	tool := mcp.NewTool("get_team_stats",
		mcp.WithDescription("Get statistics for a team"),
		mcp.WithString("team", mcp.Required(), mcp.Description("Team name")),
		mcp.WithString("competition", mcp.Description("Competition filter: brasileirao, copa, libertadores")),
		mcp.WithNumber("season", mcp.Description("Season year")),
		mcp.WithString("venue", mcp.Description("Venue filter: home, away, or both (default: both)")),
	)

	handler := func(ctx context.Context, req mcp.CallToolRequest) (*mcp.CallToolResult, error) {
		args := req.GetArguments()
		teamParam, _ := args["team"].(string)
		competitionParam, _ := args["competition"].(string)
		seasonFloat, _ := args["season"].(float64)
		season := int(seasonFloat)
		venueParam, _ := args["venue"].(string)
		if venueParam == "" {
			venueParam = "both"
		}
		venueParam = strings.ToLower(venueParam)

		type stats struct {
			Matches      int
			Wins         int
			Draws        int
			Losses       int
			GoalsFor     int
			GoalsAgainst int
		}

		var st stats
		competitions := map[string]bool{}

		for _, m := range matches {
			if competitionParam != "" && !matchesCompetition(competitionParam, m.Competition) {
				continue
			}
			if season != 0 && m.Season != season {
				continue
			}

			isHome := teamMatches(teamParam, m.HomeTeam)
			isAway := teamMatches(teamParam, m.AwayTeam)

			if !isHome && !isAway {
				continue
			}

			if venueParam == "home" && !isHome {
				continue
			}
			if venueParam == "away" && !isAway {
				continue
			}

			competitions[m.Competition] = true
			st.Matches++

			if isHome {
				st.GoalsFor += m.HomeGoal
				st.GoalsAgainst += m.AwayGoal
				if m.HomeGoal > m.AwayGoal {
					st.Wins++
				} else if m.HomeGoal == m.AwayGoal {
					st.Draws++
				} else {
					st.Losses++
				}
			} else {
				st.GoalsFor += m.AwayGoal
				st.GoalsAgainst += m.HomeGoal
				if m.AwayGoal > m.HomeGoal {
					st.Wins++
				} else if m.AwayGoal == m.HomeGoal {
					st.Draws++
				} else {
					st.Losses++
				}
			}
		}

		compList := []string{}
		for c := range competitions {
			compList = append(compList, c)
		}

		winRate := 0.0
		if st.Matches > 0 {
			winRate = float64(st.Wins) / float64(st.Matches)
		}

		resp := map[string]interface{}{
			"team":           teamParam,
			"matches":        st.Matches,
			"wins":           st.Wins,
			"draws":          st.Draws,
			"losses":         st.Losses,
			"goals_for":      st.GoalsFor,
			"goals_against":  st.GoalsAgainst,
			"goal_diff":      st.GoalsFor - st.GoalsAgainst,
			"win_rate":       winRate,
			"competitions":   compList,
		}

		b, _ := json.Marshal(resp)
		return mcp.NewToolResultText(string(b)), nil
	}

	return server.ServerTool{Tool: tool, Handler: handler}
}

File: experiments/adrianco/experiment-14/runs/language=typescript_model=sonnet_prompt=ATDD/rep1/src/tools/teams.ts
import { loadAllData } from '../data/loader.js';
import { teamsMatch } from '../data/normalizer.js';

export interface TeamStatsArgs {
  team: string;
  competition?: string;
  season?: number;
}

export function getTeamStats(args: TeamStatsArgs): any {
  const { team, competition, season } = args;
  const data = loadAllData();

  let matches: Array<{ home_team: string; away_team: string; home_goal: number; away_goal: number; season: number }> = [];

  if (!competition || competition === 'brasileirao') {
    data.brasileirao.forEach(m => {
      if (season !== undefined && Number(m.season) !== season) return;
      matches.push({ home_team: m.home_team, away_team: m.away_team, home_goal: Number(m.home_goal), away_goal: Number(m.away_goal), season: Number(m.season) });
    });
  }

  if (!competition || competition === 'copa_do_brasil') {
    data.cup.forEach(m => {
      if (season !== undefined && Number(m.season) !== season) return;
      matches.push({ home_team: m.home_team, away_team: m.away_team, home_goal: Number(m.home_goal), away_goal: Number(m.away_goal), season: Number(m.season) });
    });
  }

  if (!competition || competition === 'libertadores') {
    data.libertadores.forEach(m => {
      if (season !== undefined && Number(m.season) !== season) return;
      matches.push({ home_team: m.home_team, away_team: m.away_team, home_goal: Number(m.home_goal), away_goal: Number(m.away_goal), season: Number(m.season) });
    });
  }

  let wins = 0, losses = 0, draws = 0, goals_scored = 0, goals_conceded = 0;

  for (const m of matches) {
    const isHome = teamsMatch(m.home_team, team);
    const isAway = teamsMatch(m.away_team, team);
    if (!isHome && !isAway) continue;

    const scored = isHome ? m.home_goal : m.away_goal;
    const conceded = isHome ? m.away_goal : m.home_goal;

    goals_scored += scored;
    goals_conceded += conceded;

    if (scored > conceded) wins++;
    else if (scored < conceded) losses++;
    else draws++;
  }

  const played = wins + losses + draws;
  const points = wins * 3 + draws;

  return {
    team,
    competition: competition || 'all',
    season: season || 'all',
    matches_played: played,
    wins,
    losses,
    draws,
    goals_scored,
    goals_conceded,
    points,
  };
}
